pkg/network/node/egressip.go

```diff
@@ -0,0 +1,302 @@
+}
+
+func (eip *egressIPWatcher) watchHostSubnets() {
+	common.RunEventQueue(eip.networkClient.Network().RESTClient(), common.HostSubnets, func(delta cache.Delta) error {
+		hs := delta.Object.(*networkapi.HostSubnet)
+
+		var egressIPs []string
+		if delta.Type != cache.Deleted {
+			egressIPs = hs.EgressIPs
+		}
+
+		eip.updateNode(hs.HostIP, egressIPs)
+		return nil
+	})
+}
+
+func (eip *egressIPWatcher) updateNode(nodeIP string, nodeEgressIPs []string) {
+	eip.Lock()
+	defer eip.Unlock()
+
+	node := eip.nodesByNodeIP[nodeIP]
+	if node == nil {
+		if len(nodeEgressIPs) == 0 {
+			return
+		}
+		node = &nodeEgress{nodeIP: nodeIP, egressIPs: sets.NewString()}
+		eip.nodesByNodeIP[nodeIP] = node
+	} else if len(nodeEgressIPs) == 0 {
+		delete(eip.nodesByNodeIP, nodeIP)
+	}
+	oldEgressIPs := node.egressIPs
+	node.egressIPs = sets.NewString(nodeEgressIPs...)
+
+	// Process new EgressIPs
+	for _, ip := range node.egressIPs.Difference(oldEgressIPs).UnsortedList() {
+		eip.nodesByEgressIP[ip] = node
+		hex := ipToHex(ip)
+		claimedNodeIP := nodeIP
+
+		if nodeIP == eip.localIP {
+			if err := eip.claimEgressIP(ip, hex); err != nil {
+				glog.Errorf("Error claiming Egress IP %q: %v", ip, err)
+				claimedNodeIP = ""
+			}
+		}
+
+		if ns, exists := eip.namespacesByEgressIP[ip]; exists {
+			if ns.assignedIP == "" {
+				ns.assignedIP = ip
+				ns.nodeIP = claimedNodeIP
+				err := eip.oc.UpdateNamespaceEgressRules(ns.vnid, claimedNodeIP, hex)
+				if err != nil {
+					glog.Errorf("Error updating Namespace egress rules: %v", err)
+				}
+			}
+		}
+	}
+
+	// Process removed EgressIPs
+	for _, ip := range oldEgressIPs.Difference(node.egressIPs).UnsortedList() {
+		delete(eip.nodesByEgressIP, ip)
+		hex := ipToHex(ip)
+
+		if nodeIP == eip.localIP {
+			if err := eip.releaseEgressIP(ip, hex); err != nil {
+				glog.Errorf("Error releasing Egress IP %q: %v", ip, err)
+			}
+		}
+
+		if ns, exists := eip.namespacesByEgressIP[ip]; exists {
+			if ns.assignedIP == ip {
+				ns.assignedIP = ""
+				ns.nodeIP = ""
+				err := eip.oc.UpdateNamespaceEgressRules(ns.vnid, "", hex)
+				if err != nil {
+					glog.Errorf("Error updating Namespace egress rules: %v", err)
+				}
+			}
+		}
+	}
+}
+
+func (eip *egressIPWatcher) watchNetNamespaces() {
+	common.RunEventQueue(eip.networkClient.Network().RESTClient(), common.NetNamespaces, func(delta cache.Delta) error {
+		netns := delta.Object.(*networkapi.NetNamespace)
+
+		var egressIP string
+		if delta.Type != cache.Deleted && len(netns.EgressIPs) != 0 {
+			egressIP = netns.EgressIPs[0]
+		}
+
+		eip.updateNamespace(netns.NetID, egressIP)
+		return nil
+	})
+}
+
+func (eip *egressIPWatcher) updateNamespace(vnid uint32, egressIP string) {
+	eip.Lock()
+	defer eip.Unlock()
+
+	ns := eip.namespacesByVNID[vnid]
+	if ns == nil {
+		if egressIP == "" {
+			return
+		}
+		ns = &namespaceEgress{vnid: vnid}
+		eip.namespacesByVNID[vnid] = ns
+	}
```

Comment:
When namespace is deleted, where are we deleting vnid from namespacesByVNID map?